### ISSUE TITLE
Incorrect gradients when output is expectation value of sum of operators

### ISSUE THREAD

**Issue by ermas3** [MatchCake/MatchCake#56]
Hi,

Recently I posted an issue about incorrect outputs when calculating expectation values of sums of Pauli strings, which you kindly fixed. Thank you for this! 

Unfortunately, now I seem to be getting a similar issue again when calculating gradients. When outputting the expectation value of a single Pauli operator the gradient is correctly calculated and also matches the gradient provided by the `default.qubit()` device. However, when measuring the expectation value of a sum of Pauli operators the gradient provided by `NonInteractingFermionicDevice` is incorrect. An example of the issue is provided below, where I am running the exact same circuit on both the `default.qubit()` device and the `NonInteractingFermionicDevice` device, but getting different gradients when the output is $\braket{Z_0 + Z_1}$, whereas the gradients match when the output is just $\braket{Z_0}$ or $\braket{Z_1}$.

```
n_wires = 2

init_param = np.pi/7
init_params_nif = torch.Tensor([init_param]).requires_grad_()
init_params_std = torch.Tensor([init_param]).requires_grad_()

nif_dev = mc.NonInteractingFermionicDevice(wires=n_wires)
std_dev = qml.device("default.qubit", wires=n_wires)

@qml.qnode(nif_dev)
def objective_nif(theta):
    mc.operations.rxx.Rxx(theta[0], wires = [0, 1])
    return qml.expval(qml.PauliZ(0) + qml.PauliZ(1))
    #return qml.expval(qml.PauliZ(0))

@qml.qnode(std_dev)
def objective_std(theta):
    qml.evolve(qml.PauliX(wires = 0) @ qml.PauliX(wires = 1), theta[0])
    return qml.expval(qml.PauliZ(0) + qml.PauliZ(1))
    #return qml.expval(qml.PauliZ(0))

result_nif = objective_nif(init_params_nif)
result_nif.backward()

result_std = objective_std(init_params_std)
result_std.backward()

print(f"Pennylane output: {objective_std(init_params_std)}, MatchCake output: {objective_nif(init_params_nif)}. Outputs now match!")
print(f"Pennylane gradient: {init_params_std.grad}, Matchcake gradient: {init_params_nif.grad}. Gradients match when only measuring a single PauliZ operator.")
```
Furthermore, the given circuit (which is just a single Rxx-gate) we are expecting the expectation values to satisfy $\braket{Z_0} = \braket{Z_1} \implies \braket{Z_0 + Z_1} = 2 \braket{Z_0}$, which is indeed the case for `default.qubit`, but not the case for the Pennylane device.

On a possibly unrelated note, I am also constantly getting the warning

```
\lib\site-packages\matchcake\matchgate_parameter_sets\matchgate_params.py:285: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at aten\src\ATen/core/TensorBody.h:494.)
  grads = [getattr(getattr(self, f"{attr}"), "grad", None) for attr in self.ATTRS]
```

any time I am doing gradient calculations with the `NonInteractingFermionicDevice` device, despite the fact that the provided parameter should definitely be a leaf tensor. The error occurs regardless if the calculated gradients are correct or not. I am not sure if this is related to the issue regarding the faulty gradients or not.

**Comment by JeremieGince** [comment]
Hi,

Thank you for bringing this to my attention. I'll take a look and prioritize fixing it, as it's a very important issue.

Best,
Jérémie Gince

**Comment by JeremieGince** [comment]
Hi @ermas3 !

It's been a while, but I finally managed to tackle this issue. It was a tricky one since it was caused by multiple small errors that only occurred occasionally.

Anyway, you can verify that it works with the `dev` version of the package using the following code.


```python
import numpy as np
import pennylane as qml
import torch
import tqdm

import matchcake as mc

n_wires = 2
init_param = np.pi / 7
init_params_nif = torch.Tensor([init_param, ]).double().requires_grad_()
init_params_std = torch.Tensor([init_param, ]).double().requires_grad_()

nif_dev = mc.NonInteractingFermionicDevice(wires=n_wires)
std_dev = qml.device("default.qubit", wires=n_wires)


def circuit(theta):
    mc.operations.rxx.Rxx(theta, wires=[0, 1])
    return qml.expval(qml.PauliZ(0) + qml.PauliZ(1))
    # return qml.expval(qml.PauliZ(0))


@qml.qnode(nif_dev)
def objective_nif(theta):
    return circuit(theta)


@qml.qnode(std_dev)
def objective_std(theta):
    return circuit(theta)


def get_nif_result(theta):
    return nif_dev.execute_generator(
        [mc.operations.Rxx(theta, wires=[0, 1])],
        observable=qml.PauliZ(0) + qml.PauliZ(1),
        output_type="expval",
    )


torch.autograd.gradcheck(
    objective_nif,
    torch.Tensor([init_param]).double().requires_grad_(),
    raise_exception=True,
    check_undefined_grad=False,
)
print(f"{get_nif_result(torch.Tensor([init_param, ]).double().requires_grad_()) = }, "
      f"{objective_nif(torch.Tensor([init_param, ]).double().requires_grad_()) = }")
for p in tqdm.tqdm(np.linspace(0, 2 * np.pi, 100)):
    torch.autograd.gradcheck(
        get_nif_result,
        torch.Tensor([p, ]).double().requires_grad_(),
        raise_exception=True,
        check_undefined_grad=False,
    )
print("All checks passed!")

result_nif = objective_nif(init_params_nif)
result_nif.backward()
result_std = objective_std(init_params_std)
result_std.backward()
print(f"Pennylane output: {objective_std(init_params_std)}, MatchCake output: {objective_nif(init_params_nif)}. Outputs now match!")
print(f"Pennylane gradient: {init_params_std.grad}, Matchcake gradient: {init_params_nif.grad}.")

```


Thank you again for reporting this issue—it helped improve the package!

Have a great day!
Jérémie Gince
